lib/proxy_api/v2/features.rb

```diff
@@ -0,0 +1,16 @@
+module ProxyAPI::V2
+  class Features < ProxyAPI::Resource
+    def initialize(args)
+      @url = args[:url] + "/v2/features"
+      super args
+    end
+
+    def features
+      parse get
+    rescue RestClient::ResourceNotFound
+      raise NotImplementedError
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to detect features"))
+    end
```

Comment:
We need to be *much* better and verbose in case the proxy returns 404, users will hit this for sure. We need to think a good message and prepare a link to documentation or something with instructions if that's too much of text.

My wild guess: We are gonna ask the user to upgrade the proxy to particular version (Foreman core version?) here?

Reply:
Since I initially wrote most of this: my idea was that this is just a very low level library feature. At a higher level we should catch the NotImplementedError. There we can decide to fall back (as we currently do) or (at a later stage) warn the user. The reasoning for this is that a user might first deploy the newer foreman version and then the proxy.

However, you are right that we really prefer to have the new API be present and should think of ways of how to ensure this. Perhaps a regular job that diffs the current proxy registration vs the live values. This is especially important for more dynamic capabilities and settings.

Reply:
I meant a way that diffs the two. Not sure if this should be a scheduled task or a UI action (or both). Should it also automatically update the proxy registration or just notify the user?

Reply:
@ekohl i'm still not quite getting the reasoning behind this?  Just to convey to the user that something changed?

Reply:
I think it's important that the Foreman has a correct view of its proxies and their features + capabilities. Currently we have the installer that refreshes the features. For now that's ok, but I'd like it if we didn't have to rely on the installer that much. Given capabilities can be procs and thus change easily (probably outside of the installer too), I suspect the Foreman's view can get out of sync with reality.

Another thing that the new proxy registration protocol provides is a list of failed modules. The Foreman can report on this as a problem to the user. Previously they had to look in the proxy logs which can be obscure.

To be clear: it's not something that has to be part of this PR. More as a future follow up.

Reply:
@ekohl understood, thanks for the explanation!

Reply:
I have missed the `NotImplementedError` line this is good enough.